File: stack/tasks/isValid.js
/**
 * @param {string} s
 * @return {boolean}
 * 也是利用栈来匹配，匹配一个pop一个出去。最后栈为空则是有效的。
 *
 * Input: "()[]{}" Output: true
 */
var isValid = function(s) {
  let stack = [];
  const mapping = { ")": "(", "}": "{", "]": "[" };

  for (let i = 0; i < s.length; i++) {
    // 如果是右向的字符，取出栈顶元素进行比较
    if (s[i] in mapping) {
      const curr = stack.pop();
      // 这里 s[i] 是key ,mapping[s[i]] 获取到值，与 pop出来的栈顶元素进行比较
      if (mapping[s[i]] !== curr) {
        return false;
      }
    } else {
      // 不是右向元素，push到栈中
      stack.push(s[i]);
    }
  }
  return !stack.length;
};

File: isValid/isValid.java
/**
 * 给定一个只包含 '(', ')', '{', '}', '[' 和 ']'的字符串,
 * 判断是不是合法的, 这道题主要用到栈来实现 先进后出 
 * 思路是将 "(" "{" "[" push 到栈中，然后根据")" "}" "]" 从栈* * 中 pop顶层元素

 * Input: "()" Output: true
 * Input: "()[]{}" Output: true
 * Input: "(]" Output: false 
 * Input: "([)]" Output: false
 * Input: "{[]}" Output: true
 */

class Solution {
    public boolean isValid(String s) {
        // s 为空，或者为奇数，return false
        if (s == null || s.length() % 2 == 1) {
            return false;
        }
        
        // 定义一个 stack
        Stack<Character> stack = new Stack<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '[' || c == '{' || c == '(') {
                // 如果字符是左向的，push到栈中
                stack.push(c);
            } else {
                // 栈为空，且这个字符是右向，返回false
                if (stack.isEmpty()) {
                    return false;
                } else if(c == ']' && stack.pop() != '[') { 
                  // 匹配 "[" 与 "]" 匹配就pop顶层元素
                    return false;
                } else if (c == '}' && stack.pop() != '{') { 
                    return false;
                } else if (c == ')' && stack.pop() != '(') {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}